### PR TITLE
Simplify continuous webjob by using background service

### PR DESCRIPTION
**Pull request by SeanFeldman**
Suggestions made by David.

@davidfowl did I capture your recommendations ([1](https://github.com/Azure/azure-webjobs-sdk/issues/2153#issuecomment-484344259) and [2](https://github.com/Azure/azure-webjobs-sdk/issues/2153#issuecomment-484344629)) correctly?

### CODE REVIEW

File: AppInsightsWithWebJob/Program.cs [hunk 1]
```diff
@@ -37,14 +38,13 @@ static async Task Main(string[] args)
...
 
-            var telemetryClient = host.Services.GetService<TelemetryClient>();
-            AppDomain.CurrentDomain.UnhandledException += (sender, eventArgs) => telemetryClient.TrackException((Exception)eventArgs.ExceptionObject);
-            TaskScheduler.UnobservedTaskException += (sender, eventArgs) => telemetryClient.TrackException((Exception)eventArgs.Exception);
+//            var telemetryClient = host.Services.GetService<TelemetryClient>();
+//            AppDomain.CurrentDomain.UnhandledException += (sender, eventArgs) => telemetryClient.TrackException((Exception)eventArgs.ExceptionObject);
+//            TaskScheduler.UnobservedTaskException += (sender, eventArgs) => telemetryClient.TrackException((Exception)eventArgs.Exception);
 
-            using (cancellationToken.Register(() => host.Services.GetService<IJobHost>().StopAsync().GetAwaiter().GetResult()))
             using (host)
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs]: Remove the using. RunAsync does this already .

File: AppInsightsWithWebJob/Program.cs [hunk 2]
```diff
@@ -37,14 +38,13 @@ static async Task Main(string[] args)
...
             {
-                await host.RunAsync(cancellationToken);
+                await host.RunAsync(default(CancellationToken));
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs]: Remove the argument from RunAsync there's an overload that doesnt take an argument.

File: AppInsightsWithWebJob/Program.cs [hunk 3]
```diff
@@ -37,14 +38,13 @@ static async Task Main(string[] args)
             var host = builder.Build();
             var cancellationToken = new WebJobsShutdownWatcher().Token;
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs]: What is this?

File: AppInsightsWithWebJob/Program.cs [hunk 4]
```diff
@@ -37,14 +38,13 @@ static async Task Main(string[] args)
...
 
-            var telemetryClient = host.Services.GetService<TelemetryClient>();
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs:40]: You can log exceptions in main if RunAsync throws

File: AppInsightsWithWebJob/Program.cs (hunk 3 above)
Comment (SeanFeldman) [reply AppInsightsWithWebJob/Program.cs]: Left over from the original code. Will remove.

File: AppInsightsWithWebJob/Program.cs (hunk 4 above)
Comment (SeanFeldman) [reply AppInsightsWithWebJob/Program.cs:40]: So a simple `try/catch`? OK, I'll give that a shot.

File: AppInsightsWithWebJob/Program.cs [hunk 5]
```diff
@@ -8,6 +7,10 @@
 
 namespace AppInsightsWithWebJob
 {
+    using System;
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs:3]: nit: Fix this 😄

File: AppInsightsWithWebJob/Program.cs [hunk 6]
```diff
@@ -35,16 +38,16 @@ static async Task Main(string[] args)
...
-            using (cancellationToken.Register(() => host.Services.GetService<IJobHost>().StopAsync().GetAwaiter().GetResult()))
-            using (host)
+            try
+            {
+                await host.RunAsync();
+            }
+            catch (Exception exception)
             {
-                await host.RunAsync(cancellationToken);
+                telemetryClient.TrackException(exception);
```
Comment (davidfowl) [inline AppInsightsWithWebJob/Program.cs]: This is good for errors on startup but you also need to handle errors in the background service itself.

File: AppInsightsWithWebJob/ContinuousJob.cs [hunk 7]
```diff
@@ -15,13 +15,13 @@ public ContinuousJob(ILogger<ContinuousJob> logger)
             this.logger = logger;
         }
 
-        public async Task StartAsync(CancellationToken cancellationToken)
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
             var counter = 0;
-            while (!cancellationToken.IsCancellationRequested)
+            while (!stoppingToken.IsCancellationRequested)
```
Comment (davidfowl) [inline AppInsightsWithWebJob/ContinuousJob.cs:46]: Add the try catch in here.

File: AppInsightsWithWebJob/Program.cs (hunk 6 above)
Comment (SeanFeldman) [reply AppInsightsWithWebJob/Program.cs]: Makes sense. Doing that now.

File: AppInsightsWithWebJob/ContinuousJob.cs [hunk 8]
```diff
@@ -1,39 +1,45 @@
...
-using System.Threading;
-using System.Threading.Tasks;
-using Microsoft.Extensions.Hosting;
-using Microsoft.Extensions.Logging;
-
-namespace AppInsightsWithWebJob
+﻿namespace AppInsightsWithWebJob
 {
-    public class ContinuousJob : IHostedService
+    using System;
```
Comment (davidfowl) [inline AppInsightsWithWebJob/ContinuousJob.cs:3]: Sorry, move them out of the namespace not into it. At least thats the default style.

**Comment by davidfowl** [comment]
The other thing you should try is not directly using telemetry client and just logging the exception via ILogger. It should properly flow to AI if it is wired up properly and will make your code more portable.

File: AppInsightsWithWebJob/ContinuousJob.cs (hunk 8 above)
Comment (SeanFeldman) [reply AppInsightsWithWebJob/ContinuousJob.cs:3]: Oh, ignore that. Old habit of what we do internally.

**Comment by SeanFeldman** [comment]
I just did exactly that here: https://github.com/SeanFeldman/ApplicationInsightsWithWebJobs/pull/1/commits/af538e495ef6d22562220f220d7c592babb2bbab. Still waiting to see if it was logged in AI.

**Comment by SeanFeldman** [comment]
For some reason, with an instrumentation key specified (`x` replaced with a real value)
https://github.com/SeanFeldman/ApplicationInsightsWithWebJobs/blob/af538e495ef6d22562220f220d7c592babb2bbab/AppInsightsWithWebJob/Program.cs#L27
Applications Insights still doesn't get anything.

**Comment by SeanFeldman** [comment]
@davidfowl unrelated question - do I need to mark the host with `.UseConsoleLifetime()`?

**Comment by davidfowl** [comment]
In 3.0 we made that the default lifetime. I can’t recall the 2.2 behavior (I’ll check). If ctrl+c worked then it’s on

**Comment by SeanFeldman** [comment]
The project is on .NET Core 2.2 and <kbd>ctrl</kbd>+<kbd>c</kbd> is there. So yes, looks like `.UseConsoleLifetime()` is on by default.
image

**Comment by davidfowl** [comment]
Isn't there an API you can use in app insights to let it flush? Something like Serilog's close and flush https://merbla.com/2016/07/06/serilog-log-closeandflush/

**Comment by SeanFeldman** [comment]
Not according to [this wiki](https://github.com/Azure/azure-webjobs-sdk/wiki/Application-Insights-Integration#flushing-application-insights-logs-1):

> Flushing Application Insights Logs
> No extra action is required in v3 (unlike v2) to flush the TelemetryClient when the host stops. The .NET Core dependency injection system will automatically dispose of the registered ApplicationInsightsLoggerProvider, which will flush the TelemetryClient.

But it doesn't seem to be the case.

**Comment by SeanFeldman** [comment]
Interesting enough, log from `StopAsync` didn't make it through.

@lmolkova I've wired AppInsights, but find it interesting that not all the logs are flushed. For example, the log from `StopAsync()` is not written. Nor the logs from the `Program.Main()`.

https://github.com/SeanFeldman/ApplicationInsightsWithWebJobs/blob/de44ff38cf089ace12bf46f4d88bbcf34cab0ee9/AppInsightsWithWebJob/ContinuousJob.cs#L36-L41

https://github.com/SeanFeldman/ApplicationInsightsWithWebJobs/blob/suggestions-by-davidfowl/AppInsightsWithWebJob/Program.cs#L12

Any pointers what should be done?

**Comment by SeanFeldman** [comment]
@davidfowl I've raised an issue with WebJobs SDK. Seems like it's not playing nicely with the `BackgroundService`
https://github.com/Azure/azure-webjobs-sdk/issues/2182

**Comment by lmolkova** [comment]
> I've wired AppInsights, but find it interesting that not all the logs are flushed. For example, the log from StopAsync() is not written. Nor the logs from the Program.Main().

I cannot repro it - took your code and run it and I get all telemetry. Could you still repro it?

I believe the possible issue is some ingestion incidents recently and they mostly affected .NET Core SDKs because of https://github.com/Microsoft/ApplicationInsights-dotnet/issues/1047, https://github.com/Microsoft/ApplicationInsights-dotnet/issues/1049 and https://github.com/Microsoft/ApplicationInsights-dotnet/pull/1114.

We are going to ship stable version of ApplicationInsights SDKs with all these fixes and update web jobs sdk with it shortly.

You may also install latest preview of Microsoft.ApplicationInsights package that should have fixes.

**Comment by SeanFeldman** [comment]
> I cannot repro it - took your code and run it and I get all telemetry. Could you still repro it?

I start suspecting there are a few things on my end:
1. There's at least a few minutes delay between emitted information and analyzer returning it.
2. There's an overwhelming amount of logs sent. I've yet to identify a good way of filtering things out.  (have to admit doco is not really stellar about this topic)

Good to know about those issue. Those could influence it as well.
I'll try the preview version. Not confident using it in production though 🙂

**Comment by lmolkova** [comment]
We're going to ship stable version pretty soon (beginning  of May)

**Comment by SeanFeldman** [comment]
@lmolkova for the logging adapters, there's no preview build. Latest for NLog (Microsoft.ApplicationInsights.NLogTarget), for example, is [2.9.1](https://www.nuget.org/packages/Microsoft.ApplicationInsights.NLogTarget/2.9.1) from two months ago.

**Comment by SeanFeldman** [comment]
Unless what you meant is to force the transient dependency to be an explicit one and reference beta version of Microsoft.ApplicationInsights ([2.10.0-beta3](https://www.nuget.org/packages/Microsoft.ApplicationInsights/2.10.0-beta3)).

**Comment by lmolkova** [comment]
right, the fixes are related to the channel to the backend and it is enough to install the Microsoft.ApplicationInsights (2.10.0-beta3) and (sorry, forgot to mention before) Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel (2.10.0-beta3).

**Comment by SeanFeldman** [comment]
@lmolkova are you sure about `Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel`? This doesn't seem to be a dependency `Microsoft.Extensions.Logging.ApplicationInsights` is taking.

**Comment by lmolkova** [comment]
yes, I'm sure. this is a dependency of Microsoft.Azure.WebJobs.Logging.ApplicationInsights

**Comment by SeanFeldman** [comment]
I see what's going on. I've got confused with another project with webjobs where `Microsoft.Azure.WebJobs.Logging.ApplicationInsights` is not used. That one is using NLog adapter.

Do you know where I can find doco what packages are needed for what?

**Pull request merged**